File: openharmony/arkcompiler_runtime_core/static_core/plugins/ets/tests/interop_js/tests/interface/ts_to_ets/interface_method.cpp
/**
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include "ets_interop_js_gtest.h"

namespace ark::ets::interop::js::testing {

class EtsInterfaceTsToEtsTest : public EtsInteropTest {};

TEST_F(EtsInterfaceTsToEtsTest, check_any_type_interface_class_string)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAnyTypeInterfaceClassString"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_any_type_interface_class_int)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAnyTypeInterfaceClassInt"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_any_type_interface_class_bool)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAnyTypeInterfaceClassBool"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_any_type_interface_class_array)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAnyTypeInterfaceClassArray"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_any_type_interface_class_object)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAnyTypeInterfaceClassObject"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_interface_class_any_type_method_string)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateInterfaceClassAnyTypeMethodString"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_interface_class_any_type_method_int)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateInterfaceClassAnyTypeMethodInt"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_interface_class_any_type_method_bool)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateInterfaceClassAnyTypeMethodBool"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_interface_class_any_type_method_array)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateInterfaceClassAnyTypeMethodArray"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_interface_class_any_type_method_object)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateInterfaceClassAnyTypeMethodObject"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_union_type_interface_class_string)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkUnionTypeInterfaceClassString"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_union_type_interface_class_int)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkUnionTypeInterfaceClassInt"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_interface_class_union_type_method_string)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateInterfaceClassUnionTypeMethodString"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_interface_class_union_type_method_int)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateInterfaceClassUnionTypeMethodInt"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_subset_by_ref_interface)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkSubsetByRefInterface"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_subset_by_ref_interface_error)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkSubsetByRefInterfaceError"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_subset_by_value_class)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkSubsetByValueClass"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_subset_by_value_class_from_ts)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateSubsetByValueClassFromTs"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_instance_interface_class_union_type_method_int)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkInstanceInterfaceClassUnionTypeMethodInt"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_instance_interface_class_union_type_method_string)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkInstanceInterfaceClassUnionTypeMethodString"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_instance_subset_by_value_class_from_ts)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkInstanceSubsetByValueClassFromTs"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_tuple_type_class)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkTupleTypeClass"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_tuple_class_from_ts)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateTupleClassFromTs"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_instance_tuple_class_from_ts)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkInstanceTupleClassFromTs"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_with_optional_method_class)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkWithOptionalMethodClass"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_without_optional_method_class)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkWithoutOptionalMethodClass"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_class_with_optional_method)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateClassWithOptionalMethod"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_create_class_without_optional_method)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateClassWithoutOptionalMethod"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_with_optional_method_instance_class)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkWithOptionalMethodInstanceClass"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_without_optional_method_instance_class)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkWithoutOptionalMethodInstanceClass"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_optional_arg_with_all_args)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkOptionalArgWithAllArgs"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_optional_arg_with_one_args)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkOptionalArgWithOneArgs"));
}
// NOTE (issues 17772) fix spread operator
TEST_F(EtsInterfaceTsToEtsTest, DISABLED_check_spread_operator_arg_with_all_args)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkSpreadOperatorArgWithAllArgs"));
}
// NOTE (issues 17772) fix spread operator
TEST_F(EtsInterfaceTsToEtsTest, DISABLED_check_spread_operator_arg_with_one_args)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkSpreadOperatorArgWithOneArgs"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_without_spread_operator_arg_with_all_args)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkWithoutSpreadOperatorArgWithAllArgs"));
}

TEST_F(EtsInterfaceTsToEtsTest, check_without_spread_operator_arg_with_one_args)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkWithoutSpreadOperatorArgWithOneArgs"));
}

}  // namespace ark::ets::interop::js::testing

File: openharmony/arkcompiler_runtime_core/static_core/plugins/ets/tests/interop_js/tests/interface/ts_to_ets/interface_method.js
'use strict';
/**
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, '__esModule', { value: true });
exports.withOptionalMethodInstanceClass =
	exports.withoutOptionalMethodInstanceClass =
	exports.tupleInstanceClass =
	exports.subsetByValueInstanceClass =
	exports.unionTypeMethodInstanceClass =
	exports.createInterfaceClassTupleTypeMethodFromTs =
	exports.TupleTypeMethodClass =
	exports.optionalArgArray =
	exports.optionalArg =
	exports.createClassWithoutOptionalMethod =
	exports.createClassWithOptionalMethod =
	exports.WithoutOptionalMethodClass =
	exports.WithOptionalMethodClass =
	exports.createSubsetByValueClassFromTs =
	exports.SubsetByValueClass =
	exports.subsetByRefInterface =
	exports.createInterfaceClassUnionTypeMethod =
	exports.UnionTypeMethodClass =
	exports.createInterfaceClassAnyTypeMethod =
	exports.AnyTypeMethodClass =
	exports.tsString =
	exports.tsNumber =
		void 0;
exports.tsNumber = 1;
exports.tsString = 'string';
const AnyTypeMethodClass = /** @class */ (function () {
	function anyTypeMethodClass() {}
	anyTypeMethodClass.prototype.get = function (a) {
		return a;
	};
	return anyTypeMethodClass;
})();
exports.AnyTypeMethodClass = AnyTypeMethodClass;
function createInterfaceClassAnyTypeMethod() {
	return new AnyTypeMethodClass();
}
exports.createInterfaceClassAnyTypeMethod = createInterfaceClassAnyTypeMethod;
const UnionTypeMethodClass = /** @class */ (function () {
	function unionTypeMethodClass() {}
	unionTypeMethodClass.prototype.get = function (a) {
		return a;
	};
	return unionTypeMethodClass;
})();
exports.UnionTypeMethodClass = UnionTypeMethodClass;
function createInterfaceClassUnionTypeMethod() {
	return new AnyTypeMethodClass();
}
exports.createInterfaceClassUnionTypeMethod = createInterfaceClassUnionTypeMethod;
function subsetByRefInterface(obj) {
	return obj.get();
}
exports.subsetByRefInterface = subsetByRefInterface;
const UserClass = /** @class */ (function () {
	function userClass() {
		this.value = 1;
	}
	return userClass;
})();
const SubsetByValueClass = /** @class */ (function () {
	function subsetByValueClass() {}
	subsetByValueClass.prototype.get = function () {
		return new UserClass();
	};
	return subsetByValueClass;
})();
exports.SubsetByValueClass = SubsetByValueClass;
function createSubsetByValueClassFromTs() {
	return new SubsetByValueClass();
}
exports.createSubsetByValueClassFromTs = createSubsetByValueClassFromTs;
const WithOptionalMethodClass = /** @class */ (function () {
	function withOptionalMethodClass() {}
	withOptionalMethodClass.prototype.getNum = function () {
		return exports.tsNumber;
	};
	withOptionalMethodClass.prototype.getStr = function () {
		return exports.tsString;
	};
	return withOptionalMethodClass;
})();
exports.WithOptionalMethodClass = WithOptionalMethodClass;
const WithoutOptionalMethodClass = /** @class */ (function () {
	function withoutOptionalMethodClass() {}
	withoutOptionalMethodClass.prototype.getStr = function () {
		return exports.tsString;
	};
	return withoutOptionalMethodClass;
})();
exports.WithoutOptionalMethodClass = WithoutOptionalMethodClass;
function createClassWithOptionalMethod() {
	return new WithOptionalMethodClass();
}
exports.createClassWithOptionalMethod = createClassWithOptionalMethod;
function createClassWithoutOptionalMethod() {
	return new WithoutOptionalMethodClass();
}
exports.createClassWithoutOptionalMethod = createClassWithoutOptionalMethod;
function optionalArg(arg, optional) {
	if (optional) {
		return { with: arg, without: optional };
	}
	return { with: arg };
}
exports.optionalArg = optionalArg;
function optionalArgArray(...args) {
	const arg = [];
	for (let I = 0; I < args.length; I++) {
		arg[I] = args[I];
	}
	const withOptional = arg[0];
	const withoutOptional = arg[1];
	if (withoutOptional) {
		return { with: withOptional, without: withoutOptional };
	}
	return { with: withOptional };
}
exports.optionalArgArray = optionalArgArray;
const TupleTypeMethodClass = /** @class */ (function () {
	function tupleTypeMethodClass() {}
	tupleTypeMethodClass.prototype.get = function (arg) {
		return arg;
	};
	return tupleTypeMethodClass;
})();
exports.TupleTypeMethodClass = TupleTypeMethodClass;
function createInterfaceClassTupleTypeMethodFromTs() {
	return new TupleTypeMethodClass();
}
exports.createInterfaceClassTupleTypeMethodFromTs = createInterfaceClassTupleTypeMethodFromTs;
exports.unionTypeMethodInstanceClass = new UnionTypeMethodClass();
exports.subsetByValueInstanceClass = new SubsetByValueClass();
exports.tupleInstanceClass = new TupleTypeMethodClass();
exports.withoutOptionalMethodInstanceClass = new WithoutOptionalMethodClass();
exports.withOptionalMethodInstanceClass = new WithOptionalMethodClass();
